### ISSUE TITLE
Performance Discrepancy: gpt4all Faster than Optimized llama.cpp

### ISSUE THREAD

**Issue by MillionthOdin16** [ggerganov/llama.cpp#603]
**Expected Behavior**

I am comparing the performance of two executables: llama.cpp (current version) and the default gpt4all executable (which uses a previous version of llama.cpp). I am using the same language model for both executables, and I expect the current version of llama.cpp (which is built specifically for the hardware) to perform at least as fast as the default gpt4all executable.

**Current Behavior**

The default gpt4all executable, which uses a previous version of llama.cpp, performs significantly faster than the current version of llama.cpp. Despite building the current version of llama.cpp with hardware-specific compiler flags, it consistently performs significantly slower when using the same model as the default gpt4all executable.

**Environment and Context**

I am running the comparison on a Windows platform, using the default gpt4all executable and the current version of llama.cpp included in the gpt4all project. The version of llama.cpp is the latest available (after the compatibility with the gpt4all model).

**Steps to Reproduce**

1. Build the current version of llama.cpp with hardware-specific compiler flags.
2. Execute the llama.cpp executable using the gpt4all language model and record the performance metrics.
3. Execute the default gpt4all executable (previous version of llama.cpp) using the same language model and record the performance metrics.
4. You'll see that the gpt4all executable generates output significantly faster for any number of threads or config.

Here's some context/config when I'm doing the runs:

![image](https://user-images.githubusercontent.com/102247808/228639042-23fc9484-00a0-4cb4-8ebc-df9da1ee2e5c.png)
(left panel is latest llama.cpp, right panel is gpt4all build)

This is the older version that gpt4all uses (with some tweaks): https://github.com/zanussbaum/gpt4all.cpp

*To quickly test the difference yourself you can use the gpt4all default binaries here: https://github.com/nomic-ai/gpt4all/tree/main/chat

**Comment by MillionthOdin16** [comment]
No, that's what I'm saying. The left is optimized llama.cpp, the right is unoptimized gpt4all. The unoptimized gpt4all is significantly faster than the optimized llama. So there's something wrong.

**Comment by slaren** [comment]
Without bisecting the exact commit that introduced a performance regression it is hard to do much about it. I suspect that it happened when the code for transposed matrix multiplication was replaced with a copy.

**Comment by ggerganov** [comment]
Double check you have built `llama.cpp` in Release

**Comment by x02Sylvie** [comment]
Personally had small performance regression (13b model) over last 14 days, from 207 ms to 238 ms, but its no biggie

**Comment by xor2003** [comment]
I see significant slowness when comparing  on windows latest llama.cpp 30B with gpt4all

**Comment by dakennedyd** [comment]
I also notice performance drops on x86-64 Linux, it also uses a lot more memory than before.

**Comment by jessejohnson** [comment]
>  I am using the same language model for both executables

@MillionthOdin16 How did you do that? I couldn't get this to work yesterday. See [this issue](https://github.com/nomic-ai/gpt4all/issues/5)

**Comment by MillionthOdin16** [comment]
> >  I am using the same language model for both executables
> 
> @MillionthOdin16 How did you do that? I couldn't get this to work yesterday. See [this issue](https://github.com/nomic-ai/gpt4all/issues/5)

Did you try it in the last few hours? There were some commits couple hours ago that made it easy. Look at the readme section that was added for GPT4all

**Comment by BrunoIsaac27** [comment]
what is gpt4all and what changes to llama.cpp and alpaca.cpp? I arrived by parachute here in the conversation, I don't really know the context, I apologize

**Comment by anzz1** [comment]
We should definitely look into this as this definitely shouldn't be the case. I'm pretty confident though that enabling the optimizations didn't do that since when we did that #375 the perf was pretty well researched. If performance got lost and memory usage went up somewhere along the way, we'll need to look at where this happened. If it doesn't run well, everyone just going to fork from a older point in time instead.

**Comment by ggerganov** [comment]
Someone on Windows should try and bisect to see where this alleged degradation happens - I don't have a Windows machine. I'm pretty sure there is no performance degradation no Mac OS and Linux. The only slowness introduced, as @slaren mentioned, was the removal of the transposed `ggml_mul_mat` path which led to about %10 performance loss during **single-token inference** (i.e. after prompt ingestion). But this should have been compensated by the various updates in the SIMD code.

I guess the `ggml_mul_mat` change could have had a bigger impact on the **prompt** inference performance, but **only if you are not linking to BLAS**. In that case, the solution is to link to BLAS - you will gain significant speed-up during prompt ingestion.

But overall, we need some more specific info about your setup and timing numbers to be able pinpoint the problem.

**Comment by slaren** [comment]
>I was thinking what we'd need is a script which: start -> create directory -> git clone and build commit id -> log the performance of some runs to a file -> remove directory -> loop to start

This can be done as a binary search too, the `git bisect` command should help with that.

**Comment by anzz1** [comment]
> This can be done as a binary search too, the `git bisect` command should help with that.

Very interesting, I actually had no idea such a command existed. Seems useful in many cases. However in this case since there also was the format change somewhere in the middle, it'd simply be easier to go about it sequentially until you run to the one where the format was changed, change the model only once and proceed.

**Comment by KASR** [comment]
I've written a small python script to benchmark the token time as a function of number of threads. I've added the script in attachment if anyone want to try it. ( --> [benchmark_threads.txt](https://github.com/ggerganov/llama.cpp/files/11108972/benchmark_threads.txt) , I had to change the extension in order to upload ) It could be useful to benchmark performance for different versions. If not just ignore this message :)

Below you can see the result from my pc. I'm using windows 10, the typical system information looks like:

`system_info: n_threads = xx/ 36 | AVX = 1 | AVX2 = 1 | AVX512 = 1 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | VSX = 0 |`

I didn't go up to the 36 threads, you can see the results below. The script runs each prompt 5 times and plots the average token times. For some reason the timings go up at around 18 threads (i.e. the number of physical cores).  I will try it again later to see how robust the timings are.

Feel free to propose suggestions to make the benchmark more reliable.

![Token_time__threads](https://user-images.githubusercontent.com/22029863/228761620-93a1a2e7-6287-44b5-b6a0-f3c0182611d0.png)

**Comment by slaren** [comment]
@KASR at the moment it seems that you are only measuring the prompt eval time, I would recommend considering the prompt eval time and the regular (generation) eval time separately. You can safely assume that 1 run = 1 token in the eval time.

**Comment by anzz1** [comment]
@KASR 
That looks good, if you were to do a similar graph but where the x-axis was *commits*, 
x = 0 being https://github.com/ggerganov/llama.cpp/commit/9b4a15b17d8395eb075379b140fcd0b0283f4ef6
and x = 180 being https://github.com/ggerganov/llama.cpp/commit/9cbc404ba6699a9ba4925ea25a60552b13491c7a

It would solve our problem right then and there of figuring out where the increases and decreases happened.

**Comment by MillionthOdin16** [comment]
Wow, thanks for the critical thinking guys. You've mentioned some pretty interesting points. It's pretty crazy to see what affects performance, and looking at some of the discussions it seems like there are things like Intel perf cores they can have a significant impact (although not in my case 🙃).


I can definitely help with testing and performance metrics, I just need to make a script that'll get reliable builds between versions for my environment. It's pretty picky and often needs tweaks to make the build succeed.

One of the differences/struggles right now is that the current llamaCPP gives much more performance metric info than the build used in gpt4all. So it's hard to see the specific timings in the older gpt4all version. Apart from that, I'd want to make sure that the info I'm collecting while running builds for specific commits is actually the info that will help us.

> But overall, we need some more specific info about your setup and timing numbers to be able pinpoint the problem.

As for my build and build process, I have a Ryzen 3900x (12c, 24t) and use CMake and ninja to build my executables. I've also built with blas linked, but haven't seen a noticeable difference while using the library vs not. Other than that I use avx avx2 maxv maxv2 f16c sss3 on Release. I usually run with -t 8. And the models I use are the 4-bit quantized 7B.

> I guess the `ggml_mul_mat` change could have had a bigger impact on the **prompt inference** performance, but **only if you are not linking to BLAS**. In that case, the solution is to link to BLAS - you will gain significant speed-up during prompt ingestion.

Where should I expect to see the performance increases when I'm running with BLAS? Is it during larger completions after the prompt is loaded?

I could also do things in WSL2, but I'm not sure about the performance impacts, which is why I currently don't use it. If you think it would be better let me know.


Again, awesome job guys! You're having a huge impact on making these models accessible to normal people 🔥🔥🔥

**Comment by anzz1** [comment]
@MillionthOdin16 For Windows definitely the most common configuration (4 or 8 threads, AVX=yes, SSE3=yes, F16C=yes, AVX2=yes , AVX512=no , BLAS=no , WSL2=no) would be the best to base the benchmarks on. Obviously if you want and have the time for it, more is always better. 

The most important thing to know would be the performance data between commits starting from https://github.com/ggerganov/llama.cpp/commit/9b4a15b17d8395eb075379b140fcd0b0283f4ef6 and ending to https://github.com/ggerganov/llama.cpp/commit/9cbc404ba6699a9ba4925ea25a60552b13491c7a .

That is the thing which will help in understanding where the decreases happened. Since there has been many commits with optimizations and performance increases, it makes no sense that gpt4all/alpaca/llama-9b4a15b is faster, it should be slower because they don't have any of the recent optimizations. That leads to only one conclusion that there must have been significant decreases at some points in the timeline. It can be something not easily seen like compiler dropping inlining because of some change (inline and force-inline isn't same, and compiler can even drop force-inline) or a mistake somewhere, cannot know really. Only data can save us. :smile:

**Comment by cyyynthia** [comment]
I've been able to observe some performance degradation on Arch Linux as well. I didn't have time to look for the precise commit yet, but I found the potentially helpful information that the degradation seemed to have happened after the ggml format migration, which may help simplifying the exploration. I think it would be nice if someone else could confirm this and make sure this isn't something that happens for me only 😅

I'll keep doing some exploration, but here are the numbers I observed so far:

System info: Arch Linux - CPU: Intel Core i7 7700K. Compiled using `make`.
n_threads = 4 / 8 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | VSX = 0
All of the runs were run with the same parameters, using the 7B model (Q4_0)

I only use 4 threads, as 8 threads tend to cause performance degradation for me.

ed3c680
```
llama_print_timings:        load time =  3013.13 ms
llama_print_timings:      sample time =   277.93 ms /   512 runs   (    0.54 ms per run)
llama_print_timings: prompt eval time =  1990.05 ms /    14 tokens (  142.15 ms per token)
llama_print_timings:        eval time = 131235.49 ms /   511 runs   (  256.82 ms per run)
llama_print_timings:       total time = 137390.49 ms
```

074bea2 (first commit using new format)
```
main:     load time =  1953.12 ms
main:   sample time =   248.32 ms
main:  predict time = 102235.23 ms / 194.73 ms per token
main:    total time = 105018.24 ms
```

9b4a15b
```
main:     load time =  1385.53 ms
main:   sample time =   247.39 ms
main:  predict time = 101518.52 ms / 193.00 ms per token
main:    total time = 103729.24 ms
```

**Comment by anzz1** [comment]
Thank you @cyyynthia , something is definitely up here.

Interestingly, the new format made load times go up 40% but the sampling and predict times stayed the same (within margin of error)

I've only now woken up to this since you don't tend to see marginal changes (like in general, in anything) as I've always been on the latest version. But obviously now **everything** is much slower, loading/sampling/prompt evaluation,  and this is a high priority issue.

**Comment by cyyynthia** [comment]
After a bit more digging, #439 seems to be a very clear culprit of performance drops: eval time goes from 175ms (better than before!) @ 404e1da to 244ms @ 483bab2.

It seems other parameters does fluctuate in some intriguing ways, with increased loading times and sample times. I'll try to put together a test script and plot the evolution of these values over time on my machine.

**Comment by cyyynthia** [comment]
I've done a first test to see, and I've already gathered some interesting data. I have ran my script on 18 commits from the range cited earlier, skipping 10 commits every time. Doing it on the full range will take a while so I wanted to see what I could get without too much precision.

Here are the following graphs I've gathered, and the raw csv data if you want to explore it some more. I'll run it on all 180 commits later, probably tomorrow.

[result.csv](https://github.com/ggerganov/llama.cpp/files/11114389/result.csv)

![execution time](https://user-images.githubusercontent.com/9999055/228919179-75d5544b-52de-4661-b6b2-9d653220cda6.png)
![eval token time](https://user-images.githubusercontent.com/9999055/228919300-697e0b45-bca7-435f-a9c8-24f2890774c5.png)
![memory usage](https://user-images.githubusercontent.com/9999055/228919330-5a164c30-8fb5-405d-83af-398a1407d3f3.png)

**Comment by ggerganov** [comment]
@cyyynthia 

Thank you very much for this effort and the detailed analysis.

The memory increase is expected and easy to explain - see https://github.com/ggerganov/llama.cpp/pull/473#issuecomment-1483399313

The CPU increase is surprising. I am mostly running on Apple Silicon and it looks like I have overlooked this performance degradation. Will look further for ways to resolve it

**Comment by cyyynthia** [comment]
I figured the memory increase was expected, but I found it quite interesting nonetheless. The load times graph is also quite interesting: 

![loading times](https://user-images.githubusercontent.com/9999055/228960475-a5577b90-79f7-4eaf-8815-a5fa653804d5.png)

If anyone is interested in running the test on their machine too, here is the script I used (beware, it's written in NodeJS because I'm not very much of a Python dev and it was much faster for me to put together, and it might need slight tweaks depending on your machine 😅) Simply put the script in your llama.cpp folder (next to the Makefile) and run it.

https://gist.github.com/cyyynthia/43784451936e2a608566c42b0bacceac

**Comment by KASR** [comment]
@cyyynthia fyi I'm trying to running all the commits, I've converted the code to python and extended it a bit so that it works with windows/cmake/vs. I run each commit 5 times to obtain average timings. I will try to post some results if it finishes without errors.

I still have to figure out something: at which commit do you switch from llama.exe to main.exe?

```
LLAMA_PATH = 'C:/DATA/TestLLama/CommitTest'

LLAMA_EXE_0 = os.path.join(LLAMA_PATH, 'build/Release/llama.exe')
LLAMA_EXE_1 = os.path.join(LLAMA_PATH, 'build/bin/Release/main.exe')
```

Or did you use another trick to avoid the folder/name change?

**Comment by cyyynthia** [comment]
> Or did you use another trick to avoid the folder/name change?

I started collecting data since commit 2d64715ad475f192a4004a52d134c67ccb6f44ad, anything before is skipped and it was already named main at this time. (at least, using make)

> it got stuck after 35 commits
24568371ae0d7caf85164abe4753f36a7dba0288 and 5c19c70ba631a8f5d54feb6634e0eea178911a84 are broken commits and do not work (infinite loop, it triggered a timeout I added to my script). You'll also experience compile 3 compile errors, at f5a77a629bd0f37ae1696747633ab42a5530ec15 928480ef5b7b03d7a07e98286aebe3d8b24457d9 and ae44e23ee36c02da0e37ab508a4b473ace724f8e. I updated my script locally before running it so it would be resilient to these issues (didn't reflect that on the gist).

I also adapted my script to skip commits which were not related to ggml/llama/the example, to speed up the process - something I haven't reflected yet on my gist.

Running 5 times and getting an average is a great idea, I haven't done that for my own thing because it would take forever and my poor CPU has been suffering enough 😅  That being said, I ran it twice, and I must say some results seem to indicate it'd have been much better to do 5 runs and peak the average.

Here's my 2 runs, from 2d64715ad475f192a4004a52d134c67ccb6f44ad to ee0c40dd6de8c3c658ae43199939ef40bb1cf408 (excluding commits which didn't introduce any code change):

[result_first_run.csv](https://github.com/ggerganov/llama.cpp/files/11122751/result_first_run.csv)
[result_second_run.csv](https://github.com/ggerganov/llama.cpp/files/11122752/result_second_run.csv)

I also updated my gists with the updates I listed above (handle llama run issues, skip irrelevant commits).

**Comment by KASR** [comment]
> [2456837](https://github.com/ggerganov/llama.cpp/commit/24568371ae0d7caf85164abe4753f36a7dba0288) and [5c19c70](https://github.com/ggerganov/llama.cpp/commit/5c19c70ba631a8f5d54feb6634e0eea178911a84) are broken commits and do not work (infinite loop, it triggered a timeout I added to my script). You'll also experience compile 3 compile errors, at [f5a77a6](https://github.com/ggerganov/llama.cpp/commit/f5a77a629bd0f37ae1696747633ab42a5530ec15) [928480e](https://github.com/ggerganov/llama.cpp/commit/928480ef5b7b03d7a07e98286aebe3d8b24457d9) and [ae44e23](https://github.com/ggerganov/llama.cpp/commit/ae44e23ee36c02da0e37ab508a4b473ace724f8e). I updated my script locally before running it so it would be resilient to these issues (didn't reflect that on the gist).

Good to know! I just skipped a few commits and let the loop continu at [074bea2](https://github.com/ggerganov/llama.cpp/commit/074bea2eb1f1349a0118239c4152914aecaa1be4) since this was included in you first run. As a result, I seem to have jumped over the 2 faulty commits. 

The loop successfully passed the 3 commits where you had the compile errors, this might be an OS/compiler/etc. issue... 

So far, the loop is at 119 / 193 commits.

I will restart the loop once it complets: I had to restart a couple of time (continuing at the commit where it failed) due to some bugs in the loop. Even though I'm looping 5 times over each commit I would like to run the loop in a single run while the pc isn't doing anything else so that we have consistent timings.

**Comment by xor2003** [comment]
On Ubuntu:
For llama.cpp + llama 30B
I see very slow responces: long answer and each word is 1-2 second. (Same time the startup with mmap is extremly fast)
But for alpaca.cpp 13B - the answers are much faster

**Comment by KASR** [comment]
@anzz1 @cyyynthia @ggerganov here are the results of the run over 191 commits, as the title indicates I've evaluated each commit 10 times to obtain the average, max and min token evaluation time. You can find the full results here -->
[Benckmark_commits_results.csv](https://github.com/ggerganov/llama.cpp/files/11129255/Benckmark_commits_results.csv)

For completeness, I'm using windows 10, cmake 3.24.3, VS22. I'm using an [intel xeon w2295](https://ark.intel.com/content/www/us/en/ark/products/198011/intel-xeon-w2295-processor-24-75m-cache-3-00-ghz.html), but limited the threads to 6 to be more representative. The benchmark was done using python, each time deleting the build folder, switching commit, rebuilding. 7 commits failed to run which I excluded both from the graph and the results in the csv file.

I've used the following options for the inference:

```
-s 147852369 
-t 6 
-p 'Here is a long story about how programming came to be: '
-n 128 
-c 1024 
--top_k 40 
--top_p 0.95 
--repeat_last_n 64 
--repeat_penalty 1.1 
```



Below you can see the plot for the token timings:

![Average_token_times](https://user-images.githubusercontent.com/22029863/229285033-22967562-edf4-40a7-aa33-ba6834159f14.png)

**Comment by xor2003** [comment]
For my test on ubuntu:
1. I done a bisect on llama.cpp with 7B model. And if compare master with Alpaca.cpp it was not performance degradation.
2. I can see 30B is much slower if compare with 7B on master. Is it fine?
I have used the following command:
./main -t 7 -s 1 -m /llama/models/30B_/ggml-model-q4_0.bin -p "Count to 3."

**Comment by ggerganov** [comment]
@KASR
Very nice work!
Your investigation ends right before the latest AVX optimizations:

- 02c5b27e91a6d18cf1043d3a2d8dbc59610ac257
- 1d08882afa647c44195f4f6495a68ea455650cae

I expect after these changes, the timings to be back to normal.
Or probably after we merge #654 

@sw 
After the `mmap` changes, the `load` time is incorrect:

https://github.com/ggerganov/llama.cpp/blob/6e7801d08d81c931a5427bae46f00763e993f54a/llama.cpp#L1681-L1685

Currently, the reported load time includes not only the page faults, but also the prompt eval time. So effectively, you get the negative number since the prompt eval time has been accounted 2 times.
We have to fix this.

**Comment by KASR** [comment]
@sw I've uploaded the script that i used as a gist ( [benchmark_commits_llama_cpp.py](https://gist.github.com/KASR/474cc677a38cfe609954c90dd1e3409b) )

The timings were simply extracted from the llama printings. But indeed, when you add the printed timings there is still a bit missing. 

e.g. below: 8922.26 - ( 5603.04 + 1169.53 + 21.99 + 1836.83) = 290.87 ms

![image](https://user-images.githubusercontent.com/22029863/229346916-f716e20b-5e06-4ac7-9562-b3b75893f5ea.png)

**Comment by x02Sylvie** [comment]
https://github.com/ggerganov/llama.cpp/commit/f5a77a629bd0f37ae1696747633ab42a5530ec15 and https://github.com/ggerganov/llama.cpp/commit/483bab2e3d4a868fe679d8bb32827d2a4df214dc seem to be major culprits in my case when I locally checked before/after above mentioned commits the miliseconds.

slowdown may seem minimal with small context but as conversation grows longer until hitting max (2048 context?) the slowdown also introduced in above commits also increases in my case atleast

**Comment by MillionthOdin16** [comment]
I've noticed this as well. Tokens per second definitely drops as the
conversation grows.

**Comment by mcepok** [comment]
> [f5a77a6](https://github.com/ggerganov/llama.cpp/commit/f5a77a629bd0f37ae1696747633ab42a5530ec15) and [483bab2](https://github.com/ggerganov/llama.cpp/commit/483bab2e3d4a868fe679d8bb32827d2a4df214dc) seem to be major culprits in my case when I locally checked before/after above mentioned commits the miliseconds.
> 
> slowdown may seem minimal with small context but as conversation grows longer until hitting max (2048 context?) the slowdown also introduced in above commits also increases in my case atleast

Just to make sure, you're saying that before these commits there was no slowdown for larger contexts? Or just a less significant slowdown?

**Comment by cyyynthia** [comment]
> Tokens per second definitely drops as the conversation grows.

I've been observing this too and I got curious; so I've did some measurements to see how bad the slowdown was. I've plotted all the token times for -n 2000 (2048 ctx). To export the timings, I modified llama.cpp line 1003 and made it print to stderr the token time.

 - CPU: Intel i7 7700K (4c 8t)
 - n_threads = 4 / 8 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | VSX = 0
 - min: 139.568 ms
 - max: 1743.623 ms
 - avg: 754.408 ms

This graph is a single-run on a slightly noisy environment, so take the graph with a grain of salt. I'm definitely intrigued by the sudden bump in token time at the middle, and the acceleration of the performance degradation. I'll try to run more tests about it, and try to extract values from older versions as well.
![image](https://user-images.githubusercontent.com/9999055/229541956-53b30731-7360-4561-ad7f-c32d9f629163.png)

**Comment by mcepok** [comment]
> I'm definitely intrigued by the sudden bump in token time at the middle

That could be context swapping, see

https://github.com/ggerganov/llama.cpp/blob/437e77855a54e69c86fe03bc501f63d9a3fddb0e/examples/main/main.cpp#L246-L265

However, if I understand it correctly, it should only happen once every about `n_ctx/2` and not for every token, what the graph implies. Anyways, for your example the location should be around `1024` which fits perfectly.

**Comment by cyyynthia** [comment]
I don't think context swap is to blame. It happens once the context is full (if I understand correctly), which in my example should never happen (since I start at 13 (prompt), and generate 2000 tokens). Only once the first swap has been performed, since we "free" half of the context space, your assumption would now hold.

Even if a swap happened, my understanding is that it would generate a freeze in generation, and go back to work as usual with no impact on the token time (or if anything a slight decrease, based on the observations that the more the context fills, the slower it gets).

----

To prove this, I ran the same test with -n 1024 -c 128. I indeed did observe pauses in the generation (as expected), and here's the graph (red lines mean a context swap):
![image](https://user-images.githubusercontent.com/9999055/229563746-a7dc1462-8365-4854-959d-87c08ca9a72e.png)

Swaps occurred after 115, 178, 241, 304, 367, 430, 493, 556, 619, 682, 745, 808, 871, 934 and 997 tokens.

**Comment by mcepok** [comment]
Yes, you're right, I just realised too that my comment makes no sense. But it's interesting that context swap actually resets the increase of time per token.

**Comment by cyyynthia** [comment]
I've analyzed data for 2000 tokens on 6 different commits and established that 483bab2e3d4a868fe679d8bb32827d2a4df214dc is **definitely** a **HUGE** regression for large prompts. The run seg-faults mid-run, but the regression in token times for large inputs is extremely visible already and confirmed by 4870e455b3653f7d7769fa5772b2c90ffad088df, jumping from 415s runtime to 1673s, a 4x increase!

The latest versions do close to nothing to compensate for the massive performance degradation caused:
![image](https://user-images.githubusercontent.com/9999055/229624035-ad6c81eb-acec-4b86-8cad-2794dba48eb7.png)

You can see at the beginning, newer versions are faster (as expected with all the great work that has been going on in optimizing SIMD thingies), but quickly get unbearably slow compared to earlier versions.

Raw CSV: 
[token_times.csv](https://github.com/ggerganov/llama.cpp/files/11142400/token_times.csv)

**Comment by KASR** [comment]
@cyyynthia out of curiosity, if you use the latest commit and remove ggml_cpy from the V_trans (i.e. revert the changes from #439 ) do you then retrieve the original timings?

**Comment by cyyynthia** [comment]
@KASR I actually tried, but because ggml.c moved on with the change and removed some code paths that were now unused, I couldn't get it to work. I tried bringing back some code bits in a sort-of naive manner but I got broken output, so I assumed I've been doing something wrong and my results might not be valid.

If someone has a working patch to test, I'd love to give it a shot on my machine and compare the results, but otherwise I'm afraid I can't test that out. That being said, the commit right before the change (404e1da38ec8025707031a8027da14dc1590f952) is tested and on the graph.

**Comment by cyyynthia** [comment]
> It's faster than latest master at first glance, but I really haven't tested it thoroughly.

@sw Oh yes it is! Only 333s runtime! 107ms per token on average, this is by far the fastest run of them all!! Here is the graph (with only interesting commits):
![image](https://user-images.githubusercontent.com/9999055/229933129-64c81ef7-db3d-436f-9bb5-066c7756e81b.png)

And the raw CSV data: 
[token_times.csv](https://github.com/ggerganov/llama.cpp/files/11153610/token_times.csv)

**Comment by MillionthOdin16** [comment]
Okay, wow. Can you summarize that graph so I can make sure I'm interpreting it correctly? The vertical axis is time in seconds and x is token count in the context? Is this a fixed test that measures ms per token, and the difference is that great?

**Comment by cyyynthia** [comment]
The vertical axis is in microseconds, and the horizontal axis is the number of tokens. I ran main with set parameters and let it generate 2000 tokens. For each token, I logged the time it took to generate it and plotted it on the graph.

Each line is a single test on a specific commit (single run). The amount of times it takes to generate a single token grows over time as there are more and more things in the context store to care about. The yellow line represents when #439 was introduced, where we can see the time it takes to generate a single token grows much faster than before (the green line). The blue line is current master branch, and we can see the regression is still very pronounced.

The red line is sw' patch which is master with #439 (and #500) reverted. We can see the time it takes to generate a single token is much lower, and grows much slower (exactly like it used to do). We can also see there is an overall performance increase, from all the SIMD optimizations that have been done in the past days.

I hope this explains it well enough!

**Comment by MillionthOdin16** [comment]
That's awesome! I always wonder what the actual performance increase is when I see commits for with additional optimizations.

This plot is super useful for evaluating performance. I wonder if we can turn it into a script that runs for each new commit / publish as a workflow both as a performance sanity check and just because it's cool to see the progress :)

Just an idea for the future. But awesome job!!!

**Comment by ivanstepanovftw** [comment]
I did not notice any performance improvement for https://github.com/sw/llama.cpp/commit/f747a43 =.=
Slightly worse perplexity and OpenBLAS broken.

https://github.com/sw/llama.cpp/commit/f747a43: 
```
main: seed = 1680654638
llama_model_load: loading model from 'models/7B/ggml-model-q4_0.bin' - please wait ...
llama_model_load: n_vocab = 32000
llama_model_load: n_ctx   = 512
llama_model_load: n_embd  = 4096
llama_model_load: n_mult  = 256
llama_model_load: n_head  = 32
llama_model_load: n_layer = 32
llama_model_load: n_rot   = 128
llama_model_load: f16     = 2
llama_model_load: n_ff    = 11008
llama_model_load: n_parts = 1
llama_model_load: type    = 1
llama_model_load: ggml map size = 4017.70 MB
llama_model_load: ggml ctx size =  81.25 KB
llama_model_load: mem required  = 5809.78 MB (+ 1026.00 MB per state)
llama_model_load: loading tensors from 'models/7B/ggml-model-q4_0.bin'
llama_model_load: model size =  4017.27 MB / num tensors = 291
llama_init_from_file: kv self size  =  256.00 MB

system_info: n_threads = 8 / 16 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | VSX = 0 | 
perplexity : calculating perplexity over 655 chunks (335687 tokens, 512 n_ctx)
33.30 seconds per pass - ETA 6.06 hours
[1]4.5738,[2]5.1249,[3]5.9216,[4]6.6188,[5]6.7176,[6]6.7239,[7]6.9237,[8]7.0217,^C
```

---
Main:
```
28.09 seconds per pass - ETA 5.11 hours
[1]4.5619,[2]5.1787,[3]6.0491,[4]6.7494,[5]6.7852,[6]6.7668,[7]6.9845,

**Comment by cyyynthia** [comment]
> I did not notice any performance improvement for [sw@f747a43](https://github.com/sw/llama.cpp/commit/f747a43) =.=

I'm highly unsure if perplexity is a meaningful test for measuring performance. As I've illustrated in my graphs, at the beginning the TPS is fine so long only very few tokens are in the context, and only as the context fills up the degradation shows up and we get further and further away from the baseline. At 512 tokens, my data shows 2x slowdown in TPS and at 2000 tokens a 7.5x slowdown.

I think the fact OpenBLAS being broken is sort of expected, as the point of the branch was to see if the same degradation of TPS was occurring if we reverted the offending commit but kept everything else - something that my tests seem to confirm - and not to find a proper fix for the degradation yet.

**Comment by MillionthOdin16** [comment]
I see a significant improvement. My experience on windows matches cyyynthia. I didn't realize how bad it had gotten until I ran the reverted version and generation was so much faster. Thanks :)

**Comment by KASR** [comment]
@ivanstepanovftw I agree with @cyyynthia we first need to confirm the root cause of the performance degradation, right now it does appear to be #439. On a side note, it has been discussed that perplexity is improved when using blas (see #406 for more info) so it's difficult to compare the results with/without blas. 

I also agree with @sw --> `ggml_cpy` was introduced for a reason. To quote ggerganov : _"Seems like the "transposed X" branch is more efficient, but not numerically stable. Will try to see if I can resolve it and if not, I will simply remove it all together from ggml."_ and indeed before 439 the inference results differ when changing e.g. the number of threads/batch size etc.

**Comment by cyyynthia** [comment]
> this has a significant impact during interactive mode

I believe this is incorrect. It's not the -n that matters, it's how many things are in the context memory (i.e. -n_ctx and how far we are in the generation/interaction). On the revert branch, I've had significantly faster responses in interactive mode on the 13B model. I don't have data to back all this up, but I'm pretty sure the impact is the same in interactive mode.

**Comment by MillionthOdin16** [comment]
Also @cyyynthia is there a way that we can adapt your process as a way to analyze future builds? 

Do you have a script or something that you were using that could help us create something like that?

People have done some pretty cool stuff and I want to make sure it isn't lost once the issue is resolved.

**Comment by cyyynthia** [comment]
> #775 was merged pretty quickly, cyyynthia if it's not too much trouble, could you update your pretty graphs with that?

@sw The graphs aren't super pretty this time because I didn't take the time to properly close everything and had a bunch of things open in background while the test was running 😅

That being said, the regression appears to be gone. 🎉  Here's the graphs and the raw CSV:
![image](https://user-images.githubusercontent.com/9999055/230187014-bf645bff-2c34-4799-8376-01516802bd9c.png)
[token_times.csv](https://github.com/ggerganov/llama.cpp/files/11162466/token_times.csv)

> Do you have a script or something that you were using that could help us create something like that?

@MillionthOdin16 As described earlier, my process was simply to shoving a `fprintf(stderr, "debug_token_time: %ld\n", ggml_time_us() - t_start_us);` in the `llama_eval_internal` function, and dumping the numbers it spat in a CSV. Nothing too complex 😅

It could be added to llama.cpp itself behind a compile flag for enabling it more easily. But it's not rocket science, and I didn't use any special script for it.

**Comment by MillionthOdin16** [comment]
Okay, thanks for summarizing the process.

That graph makes me so happy ❤️
